sycl/test-e2e/Basic/sycl_2020_images/host_unsampled_image_read.cpp

```diff
@@ -0,0 +1,102 @@
+// RUN: %{build} -o %t.out
+// TODO: Consider moving to sycl/test as this is device-independent.
+// RUN: %{run-unfiltered-devices} %t.out
+
+// Tests for direct read of unsampled_image using host accessors.
+
+#include "common.hpp"
+
+constexpr size_t IMAGE_WIDTH = 5;
+constexpr size_t IMAGE_HEIGHT = 4;
+constexpr size_t IMAGE_DEPTH = 2;
+
+constexpr size_t IMAGE_PITCH_WIDTH = 7;
+constexpr size_t IMAGE_PITCH_HEIGHT = 5 * IMAGE_PITCH_WIDTH;
+
+template <image_format Format, int Dims>
+bool checkUnsampledImageHostReadDirect(
+    unsampled_image<Dims> &Image,
+    typename FormatTraits<Format>::rep_elem_type *RefData) {
+  using PixelType = typename FormatTraits<Format>::pixel_type;
+  constexpr ImageType ImgType = ImageType::Unsampled;
+
+  host_unsampled_image_accessor<PixelType, Dims, access_mode::read> Acc(Image);
+  assert(Image.size() == Acc.size());
+
+  range<Dims> ImageRange = Image.get_range();
+  range<2> ImagePitch = getElementWisePitch<Format, Dims>(Image);
+
+  bool success = true;
+  for (size_t I = 0; I < Acc.size(); ++I) {
+    auto Coord = DelinearizeToCoord<ImgType>(I, ImageRange);
+
+    // Read the coordinate through the accessor and read the corresponding value
+    // in the reference memory.
+    PixelType ReadVal = Acc.read(Coord);
+    PixelType ExpectedVal =
+        SimulateRead<Format, ImgType>(RefData, Coord, ImagePitch, ImageRange);
+    if (!ApproxEq(ReadVal, ExpectedVal)) {
```

Comment:
For an unsampled read, should the value read be exactly the value of the simulated read?

Reply:
Right! That should be fine for unsampled read. Sadly unsampled write can have some precision errors for normalized images due to the multiple conversions, so we can only do it for read. Good shout!